File: MiniDB_Projects/Team_QuorumDB/tests/test_parser.py
"""Tests for the SQL tokenizer and parser."""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

import pytest

from minidb.sql import ast
from minidb.sql.parser import ParseError, parse, Parser


def test_create_table_with_pk_and_not_null():
    s = parse("CREATE TABLE users (id INT PRIMARY KEY, name TEXT NOT NULL, age INT)")
    assert isinstance(s, ast.CreateTable)
    assert s.name == "users"
    assert s.pk_column == "id"
    assert [c.name for c in s.columns] == ["id", "name", "age"]
    assert s.columns[1].not_null is True
    assert s.columns[0].type == "INT"


def test_create_index_unique():
    s = parse("CREATE UNIQUE INDEX ON users (name)")
    assert isinstance(s, ast.CreateIndex)
    assert (s.table, s.column, s.unique) == ("users", "name", True)


def test_insert_multiple_rows():
    s = parse("INSERT INTO users (id, name) VALUES (1, 'alice'), (2, 'bob')")
    assert isinstance(s, ast.Insert)
    assert s.columns == ["id", "name"]
    assert s.rows == [[1, "alice"], [2, "bob"]]


def test_insert_negative_and_float_and_bool():
    s = parse("INSERT INTO t VALUES (-5, 3.14, TRUE, NULL)")
    assert s.rows == [[-5, 3.14, True, None]]


def test_select_star_with_where_and():
    s = parse("SELECT * FROM users WHERE age >= 18 AND name = 'alice'")
    assert s.columns == ["*"]
    assert s.from_table == "users"
    assert len(s.where.comparisons) == 2
    c0 = s.where.comparisons[0]
    assert (c0.left.name, c0.op, c0.right.value) == ("age", ">=", 18)


def test_select_join():
    s = parse("SELECT u.name, o.total FROM users u JOIN orders o ON u.id = o.uid "
              "WHERE o.total > 100")
    assert s.from_table == "users" and s.from_alias == "u"
    assert s.columns == ["u.name", "o.total"]
    assert len(s.joins) == 1
    j = s.joins[0]
    assert j.table == "orders"
    assert j.on.is_join_predicate
    assert (j.on.left.table, j.on.left.name) == ("u", "id")


def test_delete_with_where():
    s = parse("DELETE FROM users WHERE id = 7")
    assert isinstance(s, ast.Delete)
    assert s.where.comparisons[0].right.value == 7


def test_txn_control():
    assert isinstance(parse("BEGIN"), ast.Begin)
    assert isinstance(parse("COMMIT"), ast.Commit)
    assert isinstance(parse("ROLLBACK"), ast.Rollback)


def test_script_split():
    stmts = Parser.parse_script("BEGIN; INSERT INTO t VALUES (1); COMMIT;")
    assert [type(s).__name__ for s in stmts] == ["Begin", "Insert", "Commit"]


def test_errors():
    with pytest.raises(ParseError):
        parse("SELECT FROM users")  # missing columns
    with pytest.raises(ParseError):
        parse("INSERT INTO t VALUES (1")  # unbalanced paren

File: MiniDB_Projects/Team_minions/tests/test_parser.cpp
// Tests for the SQL lexer + parser.
#include "minidb/exceptions.h"
#include "minidb/query/parser.h"
#include "test_framework.h"

using namespace minidb;

TEST(parser, create_table) {
    Statement s = Parser::parse(
        "CREATE TABLE users (id INT PRIMARY KEY, name TEXT, age INT)");
    CHECK(s.type == StmtType::CREATE_TABLE);
    CHECK_EQ(s.create_table.table, std::string("users"));
    CHECK_EQ(s.create_table.columns.size(), (size_t)3);
    CHECK_EQ(s.create_table.columns[0].name, std::string("id"));
    CHECK(s.create_table.columns[0].type == Type::INT);
    CHECK(s.create_table.columns[0].primary_key);
    CHECK(s.create_table.columns[1].type == Type::TEXT);
    CHECK(!s.create_table.columns[1].primary_key);
}

TEST(parser, create_index) {
    Statement s = Parser::parse("CREATE INDEX idx_name ON users (name)");
    CHECK(s.type == StmtType::CREATE_INDEX);
    CHECK_EQ(s.create_index.index_name, std::string("idx_name"));
    CHECK_EQ(s.create_index.table, std::string("users"));
    CHECK_EQ(s.create_index.column, std::string("name"));
}

TEST(parser, insert_single_and_multi) {
    Statement s = Parser::parse(
        "INSERT INTO users (id, name) VALUES (1, 'alice'), (2, 'bob')");
    CHECK(s.type == StmtType::INSERT);
    CHECK_EQ(s.insert.columns.size(), (size_t)2);
    CHECK_EQ(s.insert.rows.size(), (size_t)2);
    CHECK_EQ(s.insert.rows[0][0].as_int(), (int64_t)1);
    CHECK_EQ(s.insert.rows[0][1].as_text(), std::string("alice"));
    CHECK_EQ(s.insert.rows[1][1].as_text(), std::string("bob"));
}

TEST(parser, insert_negative_number) {
    Statement s = Parser::parse("INSERT INTO t VALUES (-42, 'x')");
    CHECK_EQ(s.insert.rows[0][0].as_int(), (int64_t)-42);
}

TEST(parser, select_star_where) {
    Statement s = Parser::parse("SELECT * FROM users WHERE age >= 18");
    CHECK(s.type == StmtType::SELECT);
    CHECK(s.select.select_star);
    CHECK_EQ(s.select.from_table, std::string("users"));
    CHECK_EQ(s.select.where.size(), (size_t)1);
    CHECK(s.select.where[0].op == CompOp::GE);
    CHECK_EQ(s.select.where[0].left.col, std::string("age"));
    CHECK(!s.select.where[0].right_is_column);
    CHECK_EQ(s.select.where[0].right_value.as_int(), (int64_t)18);
}

TEST(parser, select_columns_and_and) {
    Statement s = Parser::parse(
        "SELECT id, name FROM users WHERE age > 20 AND name = 'bob'");
    CHECK(!s.select.select_star);
    CHECK_EQ(s.select.columns.size(), (size_t)2);
    CHECK_EQ(s.select.where.size(), (size_t)2);
    CHECK(s.select.where[1].op == CompOp::EQ);
    CHECK_EQ(s.select.where[1].right_value.as_text(), std::string("bob"));
}

TEST(parser, select_join) {
    Statement s = Parser::parse(
        "SELECT u.name, o.item FROM users u JOIN orders o ON u.id = o.uid "
        "WHERE o.item = 'book'");
    CHECK_EQ(s.select.from_table, std::string("users"));
    CHECK_EQ(s.select.from_alias, std::string("u"));
    CHECK_EQ(s.select.joins.size(), (size_t)1);
    CHECK_EQ(s.select.joins[0].table, std::string("orders"));
    CHECK_EQ(s.select.joins[0].alias, std::string("o"));
    CHECK(s.select.joins[0].on.right_is_column);
    CHECK_EQ(s.select.joins[0].on.left.table, std::string("u"));
    CHECK_EQ(s.select.joins[0].on.right_col.table, std::string("o"));
    CHECK_EQ(s.select.columns[0].table, std::string("u"));
}

TEST(parser, delete_with_where) {
    Statement s = Parser::parse("DELETE FROM users WHERE id = 5");
    CHECK(s.type == StmtType::DELETE);
    CHECK_EQ(s.del.table, std::string("users"));
    CHECK_EQ(s.del.where.size(), (size_t)1);
    CHECK_EQ(s.del.where[0].right_value.as_int(), (int64_t)5);
}

TEST(parser, txn_control) {
    CHECK(Parser::parse("BEGIN").type == StmtType::BEGIN);
    CHECK(Parser::parse("BEGIN TRANSACTION;").type == StmtType::BEGIN);
    CHECK(Parser::parse("COMMIT;").type == StmtType::COMMIT);
    CHECK(Parser::parse("ABORT").type == StmtType::ABORT);
    CHECK(Parser::parse("ROLLBACK").type == StmtType::ABORT);
}

TEST(parser, case_insensitive_keywords) {
    Statement s = Parser::parse("select * from users where id = 1");
    CHECK(s.type == StmtType::SELECT);
    CHECK(s.select.select_star);
}

TEST(parser, errors) {
    CHECK_THROWS(Parser::parse("SELECT FROM users"));      // no select list
    CHECK_THROWS(Parser::parse("CREATE TABLE t (id FOO)"));  // bad type
    CHECK_THROWS(Parser::parse("INSERT INTO t VALUES"));     // no values
    CHECK_THROWS(Parser::parse("SELECT * FROM"));            // no table
    CHECK_THROWS(Parser::parse("garbage"));                  // unknown stmt
}
